### DISCUSSION TITLE
Grad

### DISCUSSION THREAD

**Discussion by Marcode10** [google/flax#579] [Q&A] [answered]
Is it possible to calculate the gradient of a model w.r.t input?

**Comment by jheek** [comment, marked as answer]
It by defualt `jax.grad` only computes a grad w.r.t. the first argument so you get:
```
def loss_fn(model, x):
  ....

model_grad = jax.grad(model, batch)
```

but you can also do the following:

```
model_grad, input_grad = jax.grad(model, batch, argnums=(0, 1))
```

**Reply by Marcode10** [reply]
I try to define a model and calculate its gradient by following. 

```
class LeNET(flax.nn.Module):
    def apply(self, x):
        x = flax.nn.Conv(x, features=6, kernel_size=(3, 3))
        ....
        x = flax.nn.log_softmax(x)
        return x

_, initial_params = LeNET.init_by_shape(jax.random.PRNGKey(0),[((1, 32, 32, 1), jnp.float32)])
model = flax.nn.Model(LeNET, initial_params)
```
when I calculate the gradient of a model w.r.t to some random input. 

```
x = jnp.ones((1, 32,32,1))
model_grad = jax.grad(model)(x)
```
I get the following error 
`
TypeError: Gradient only defined for scalar-output functions. Output had shape: (1, 10).
`

Can you please direct me in this example to the way of calculating the gradient of model w.r.t input, at some input?

**Reply by jheek** [reply]
The problem here is that you have multiple outputs. A gradient is only defined for scalars. So typically you compute the gradient wrt to some loss function

**Reply by Marcode10** [reply]
Yeah, I see your point. Do you advise any gradient calculator function for vector-output functions in JAX?

**Reply by jheek** [reply]
You can use `jax.jacrev` or `jax.jacfwd` to compute the Jacobian instead
